crypto/fipsmodule/ec/ec_nistp.c

```diff
@@ -0,0 +1,111 @@
+//
+// Matrix of what has been done so far:
+// 
+// | op | P-521 | P-384 | P-256 |
+// |----------------------------|
+// | 1. |   x   |   x   |   x*  |
+// | 2. |       |       |       |
+// | 3. |       |       |       |
+// | 4. |       |       |       |
+// | 5. |       |       |       |
+//  * For P-256, only the Fiat-crypto implementation in p256.c is replaced. 
+
+#include "ec_nistp.h"
+
+// Some of the functions below need temporary field element variables.
+// To avoid dynamic allocation we define nistp_felem type to have the maximum
+// size possible (which is currently P-521 curve). The values are hard-coded
+// for the moment, this will be fixed when we migrate the whole P-521
+// implementation to ec_nistp.c.
+#if defined(EC_NISTP_USE_64BIT_LIMB)
+#define NISTP_FELEM_MAX_NUM_OF_LIMBS (9)
+#else
+#define NISTP_FELEM_MAX_NUM_OF_LIMBS (19)
+#endif
+typedef ec_nistp_felem_limb ec_nistp_felem[NISTP_FELEM_MAX_NUM_OF_LIMBS];
+
+// Group operations
+// ----------------
+//
+// Building on top of the field operations we have the operations on the
+// elliptic curve group itself. Points on the curve are represented in Jacobian
+// coordinates.
+//
+// ec_nistp_point_double calculates 2*(x_in, y_in, z_in)
+//
+// The method is taken from:
+//   http://hyperelliptic.org/EFD/g1p/auto-shortw-jacobian-3.html#doubling-dbl-2001-b
+//
+// Coq transcription and correctness proof:
+// <https://github.com/mit-plv/fiat-crypto/blob/79f8b5f39ed609339f0233098dee1a3c4e6b3080/src/Curves/Weierstrass/Jacobian.v#L93>
+// <https://github.com/mit-plv/fiat-crypto/blob/79f8b5f39ed609339f0233098dee1a3c4e6b3080/src/Curves/Weierstrass/Jacobian.v#L201>
+// Outputs can equal corresponding inputs, i.e., x_out == x_in is allowed;
+// while x_out == y_in is not (maybe this works, but it's not tested).
+void ec_nistp_point_double(const ec_nistp_felem_meth *ctx,
+                           ec_nistp_felem_limb *x_out,
+                           ec_nistp_felem_limb *y_out,
+                           ec_nistp_felem_limb *z_out,
+                           const ec_nistp_felem_limb *x_in,
+                           const ec_nistp_felem_limb *y_in,
+                           const ec_nistp_felem_limb *z_in) {
+  ec_nistp_felem delta, gamma, beta, ftmp, ftmp2, tmptmp, alpha, fourbeta;
+  // delta = z^2
+  ctx->sqr(delta, z_in);
+  // gamma = y^2
+  ctx->sqr(gamma, y_in);
+  // beta = x*gamma
+  ctx->mul(beta, x_in, gamma);
+
+  // alpha = 3*(x-delta)*(x+delta)
+  ctx->sub(ftmp, x_in, delta);
+  ctx->add(ftmp2, x_in, delta);
+
+  ctx->add(tmptmp, ftmp2, ftmp2);
+  ctx->add(ftmp2, ftmp2, tmptmp);
+  ctx->mul(alpha, ftmp, ftmp2);
+
+  // x' = alpha^2 - 8*beta
+  ctx->sqr(x_out, alpha);
+  ctx->add(fourbeta, beta, beta);
+  ctx->add(fourbeta, fourbeta, fourbeta);
+  ctx->add(tmptmp, fourbeta, fourbeta);
+  ctx->sub(x_out, x_out, tmptmp);
+
+  // z' = (y + z)^2 - gamma - delta
+  // The following calculation differs from the Coq proof cited above.
```

Comment:
okay but then we can't cite the Coq proof? Because it proves something different.

Reply:
Or is there another proof for this variant?

Reply:
you are right, I clarified this in the code comment, let me know if that's fine with you.